src/cpp-common/include/cpp-common/HashMap.tpp

```diff
@@ -0,0 +1,205 @@
+}
+
+template <typename Key, typename MappedType, uint16_t maxSize>
+bool HashMap<Key, MappedType, maxSize>::upsert(const std::pair<Key, MappedType>& item) {
+    bool loopedOnce = false;
+    uint16_t index = hash(item.first);
+    do {
+        if (m_usedSpacesFlag[index] &&
+            reinterpret_cast<const std::pair<Key, MappedType>&>(m_storage[index]).first ==
+                item.first) {
+            new (&m_storage[index]) std::pair<Key, MappedType>(item);
+            return true;
+        }
+        index++;
+        // Only loop across array once
+        if (!loopedOnce && index == maxSize) {
+            loopedOnce = true;
+            index = 0;
+        }
+    } while (index < maxSize);
+    // Failed to update, trying insert
+    return insert(item);
+}
+
+template <typename Key, typename MappedType, uint16_t maxSize>
+bool HashMap<Key, MappedType, maxSize>::remove(Key key) {
+    uint16_t index = hash(key);
+    bool loopedOnce = false;
+    do {
+        if (m_usedSpacesFlag[index] == true &&
+            reinterpret_cast<const std::pair<Key, MappedType>&>(m_storage[index]).first == key) {
+            reinterpret_cast<const std::pair<Key, MappedType>&>(m_storage[index])
+                .second.~MappedType();
+            m_usedSpacesFlag[index] = false;
+            m_usedSpaces--;
+            return true;
+        }
+        index++;
+        // Only loop across array once
+        if (!loopedOnce && index == maxSize) {
+            index = 0;
+            loopedOnce = true;
+        }
+    } while (index < maxSize);
+    return false;
+}
+
+template <typename Key, typename MappedType, uint16_t maxSize>
+void HashMap<Key, MappedType, maxSize>::clear() {
+    for (size_t i = 0; i < maxSize; i++) {
+        // Could forego calling constructor since using placement new
+        if (m_usedSpacesFlag[i] == true) {
+            reinterpret_cast<std::pair<Key, MappedType>*>(&m_storage[i])->second.~MappedType();
+            m_usedSpacesFlag[i] = false;
+        }
+    }
+    m_usedSpaces = 0;
+}
+
+template <typename Key, typename MappedType, uint16_t maxSize>
+bool HashMap<Key, MappedType, maxSize>::get(Key k, MappedType& item) const {
+    uint16_t index = hash(k);
+    bool loopedOnce = false;
+    do {
+        if (m_usedSpacesFlag[index] == true) {
+            auto& pair = reinterpret_cast<const std::pair<Key, MappedType>&>(m_storage[index]);
+            if (pair.first == k) {
+                // emplace on reference passed
+                new (&item) MappedType(pair.second);
+                return true;
+            }
+        }
+        index++;
+        // Only loop across array once
+        if (!loopedOnce && index == maxSize) {
+            index = 0;
+            loopedOnce = true;
+        }
+    } while (index < maxSize);
+    return false;
+}
+template <typename Key, typename MappedType, uint16_t maxSize>
+std::optional<std::reference_wrapper<MappedType>> HashMap<Key, MappedType, maxSize>::at(Key key) {
```

Comment:
You could just do:

```cpp
std::optional<std::reference_wrapper<MappedType>> HashMap<Key, MappedType, maxSize>::at(Key key) {
       auto obj = const_cast<const Foo*>(this)->at(key); // Using the const function
       return const_cast<std::optional<std::reference_wrapper<MappedType>>>(obj); // Remobing the const of the object
}
```

Instead of duplication the code